### PR TITLE
feat(comments): adds read-only state to comments.

### PR DESCRIPTION
**Pull request by pedrobonamin**
### Description

- Creates a new context into comments plugin to resolve upsell necessary data and behaviour. 
- Adds new export `DescriptionSerializer` from `sanity` package for reusability in comments plugin.

After free trial, it adds read only state to comments, showing the comments in a disabled state and only allowing to:
- Resolve existing comments.
- Delete existing comments

It removes the ability to:
- Add new comments.
- Reply to existing comments.
- React to comments
- Edit existing comments.
- View resolved comments.

Introduces new upsell UI into the comments inspector panel.

### What to review
Comments functionality, nothing should change for enabled plans.

### CODE REVIEW

File: packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx
Comment (pedrobonamin) [inline packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx:162]: It will open the comment inspector if for some reason the upsell data hasn't loaded yet to avoid stale clicks on the comments icon. 
It should not allow to add new comments when it's read only.

File: packages/sanity/src/structure/comments/src/components/upsell/UpsellDialog.tsx
Comment (pedrobonamin) [inline packages/sanity/src/structure/comments/src/components/upsell/UpsellDialog.tsx:33]: Need to confirm with lars the width of the dialog

File: packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts
Comment (pedrobonamin) [inline packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts:30]: This should be restored before releasing the changes

File: packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializer.tsx
Comment (hermanwikner) [inline packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializer.tsx:45]: Should we access the font weight value from the theme?

```ts
const SemiboldSpan = styled.span(({theme}) => {
  const {weights} = theme.sanity.fonts.text

  return css`
    font-weight: ${weights.semibold};
  `
})
```

File: packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializer.tsx
Comment (hermanwikner) [inline packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializer.tsx:218]: Since this serializer is intended to be used for upsell blocks – should the name of the component be more specific to better reflects its purpose? Maybe something like `UpsellDescriptionSerializer`

File: packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx:35]: Since we don't render the `CommentFieldInner` component if `enabled` is `false` – we won't show any comment field action on the fields if the enabled value is: `{enabled: false, reason: 'upsell'}`. 

I think it would be nice to create a separate field component that is used for upsell. Something like:

```ts
export function CommentsField(props: FieldProps) {
  const {enabled, reason} = useCommentsEnabled()

  // Comments upsell field component
  if (!enabled && reason === 'upsell') {
    return <CommentsFieldUpsell {...props} />
  }

  // Comments enabled field component
  if (enabled) {
    return <CommentFieldInner {...props} />
  }

  // Default field component without comments
  return props.renderDefault(props)
}
```

That is, the `CommentsFieldInnerUpsell` handles opening the inspector / dialog.

File: packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx:80]: It would be nice to avoid using this hook in here and only render this component when comments is enabled. I think making creating a new `CommentsFieldUpsell` component (see earlier comment) would be nice and separate the concerns of the enabled field and the upsell field logic.

File: packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx:44]: Do we need to check if `reason` is 'upsell' here as well? Otherwise, I think we will return `null` even though we want to show the inspector with the upsell UI since we only check the `enabled` value.

File: packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx:58]: Since we use this hook in the parent component (`CommentsInspector`) – can we remove this hook here and pass the "mode" value as a prop instead?

```tsx
<RootLayer>
  <CommentsInspectorInner {...props} mode={mode} />
</RootLayer>
```

File: packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts:29]: This value needs to include the `featureEnabled` value as well. Currently, it only checks the enabled value resolved from the config API.

File: packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx:19]: Should this value be replaced with a value from the theme?

File: packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx:13]: Should this value be replaced with a value from the theme?

File: packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx:51]: Do we want to memoize this value?

File: packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx:49]: Nitpick: alphabetize props

File: packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx:64]: Nitpick: alphabetize props

File: packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx:139]: Nitpick: alphabetize props

File: packages/sanity/src/structure/comments/src/components/reactions/CommentReactionsMenuButton.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/components/reactions/CommentReactionsMenuButton.tsx:82]: Nitpick: alphabetize props

File: packages/sanity/src/structure/comments/src/context/comments/CommentsProvider.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/context/comments/CommentsProvider.tsx:54]: I'd like to keep the `CommentsProvider` not dependent on other contexts. I think it would be better to move this logic (in `handleSetStatus`) to the `handleChangeView` function in the `CommentsInspector` instead.

File: packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts]: This is intentional to support disabling comments from config.

You could have 3 type of values here:
```
 {enabled:false, reason: null}  //  Disabled from config, won't show anything related to comments.- Will remove reason from here to make it simpler to follow.
 {enabled: true, reason: null}  // Current UI, no changes
 {enabled: true, reason:  "upsell} // Shows upsell UI.
```

Pushed an update here to clarify.

File: packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/src/components/list/CommentThreadLayout.tsx:49]: Updated, but I think we should aim to make it part of the lint rules if we want to keep it 100% consistent everywhere.

File: packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/src/components/upsell/CommentsUpsellDialog.tsx:51]: I don't see anything in the parent components or in this component that will produce unwanted rerenders and would benefit from memoizing this value, given it's not an expensive to calculate. But would love to hear your thoughts on why you consider we shuold.
This is a great post that gave me guidelines to decide to use or not a memo value, https://kentcdodds.com/blog/usememo-and-usecallback#how-is-usememo-different-but-similar .

File: packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializerUtils.ts
Comment (pedrobonamin) [reply packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializerUtils.ts:1]: Yep! Thanks

File: packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializer.tsx
Comment (pedrobonamin) [reply packages/sanity/src/core/studio/components/navbar/free-trial/DescriptionSerializer.tsx:222]: Yeap, sounds good, I moved it to a reusable location instead of navbar and renamed it.

File: packages/sanity/src/structure/comments/src/context/comments/CommentsProvider.tsx
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/src/context/comments/CommentsProvider.tsx:54]: I'm not 100% sure about this, this context it's already depending on `useCommentsSetup`, it can be moved to `commentsInspector` but if later we add another button or any other async action that sets the `resolved` states we will need to duplicate the logic added here.

File: packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx:45]: enabled will be true when reason is `upsell` .

File: packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx:35]: It will require to duplicate all the logic already added into the CommentsInput component, as the only thing that changes is the handleClick action when the input has no comments.
If it has comments, it does the same as before (opens the inspector and scrolls into view).

Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/plugin/field/CommentsField.tsx:35]: however, I could pass the value to the <CommentsFieldInner> by props if we want to avoid using the `useCommentsEnabled` hook in it

File: packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts
Comment (pedrobonamin) [inline packages/sanity/src/structure/comments/src/hooks/useResolveCommentsEnabled.ts:35]: This needs to be restored before merging

File: packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx:107]: When I test this, it seems like the telemetry event won't trigger when navigating to a comment from a URL. I think this is due to the `selectedPath` state being set in the `useEffect` at line 314 when navigating from a URL, which happens after this particular effect has already executed. That is, the `selectedPath` state is not yet available when this effect runs when navigating from a URL. I think this can be solved by moving the link telemetry event inside the effect at line 314.

What do you think of moving the `panelDismissed` event to the `handleCloseInspector` function instead? This because effects is running twice in strict mode, so that way we can avoid that the dismissed event is triggered on mount. The downside of this is that we won't trigger any telemetry event when dismissing the inspector from the button in the document panel header. However, there is a `onClose` function available that can be passed to [the inspector definition](https://github.com/sanity-io/sanity/blob/next/packages/sanity/src/structure/comments/plugin/inspector/index.ts#L20) – maybe that one can be used to trigger the event?

Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/plugin/inspector/CommentsInspector.tsx:107]: re url not working:
- Did a change, will check for the `commentIdParamRef.current` instead, which is available while loading the comments. It's weird it initially worked as I expected. But thanks for teh catch.

re document_action
- Those names are defined by the growth team so I think we need to stick to it, regarding changing it o onOpen I tried it but we don't have access to the upsell context to determine if we should trigger the event or not. 

re on dismiss:
-  Tried the onClose option, it also has an onOpen, but we don't have access to the upsell context to determine if we should trigger the event or not.

File: packages/sanity/src/core/studio/upsell/upsellDescriptionSerializer/helpers.ts
Comment (drewlyton) [inline packages/sanity/src/core/studio/upsell/upsellDescriptionSerializer/helpers.ts:17]: Do these array accesses need to be checked/guarded in some way? Just wondering if accessing [-1] will ever happen/be a problem or if accessing [length+1] will ever throw an error

File: packages/sanity/src/structure/comments/src/context/upsell/CommentsUpsellProvider.tsx
Comment (hermanwikner) [inline packages/sanity/src/structure/comments/src/context/upsell/CommentsUpsellProvider.tsx:141]: What do you think of triggering this telemetry event like below instead of in an effect?

```ts
  const handleOpenDialog = useCallback(() => {
    setUpsellDialogOpen(true)

    telemetry.log(UpsellDialogViewed, {
      feature: FEATURE,
      type: 'modal',
      source: 'field_action',
    })
  }, [telemetry])
  ```

**Review by hermanwikner — APPROVED** [review]
Thank you @pedrobonamin!

File: packages/sanity/src/structure/comments/src/hooks/useCommentsUpsell.ts
Comment (drewlyton) [inline packages/sanity/src/structure/comments/src/hooks/useCommentsUpsell.ts:13]: Might be nice to add a `console.error` here to notify future devs in case the provider moves or some other implementation causes this to break

**Review by drewlyton — APPROVED** [review]
Telemetry stuff looks fine to me (although I am new so take with grain of 🧂) - left some general questions/comments. Might want to wait for @svirs's review too before merging

File: packages/sanity/src/core/studio/upsell/upsellDescriptionSerializer/helpers.ts
Comment (pedrobonamin) [reply packages/sanity/src/core/studio/upsell/upsellDescriptionSerializer/helpers.ts:17]: The upper check `isPortableTextTextBlock` is checking for children to be defined and to be of type array. 
And later when accessing the children `block.children[idx + 1]?._type` will work as a safeguard in case it's trying to access to an undefined children

File: packages/sanity/src/structure/comments/src/hooks/useCommentsUpsell.ts
Comment (pedrobonamin) [reply packages/sanity/src/structure/comments/src/hooks/useCommentsUpsell.ts:13]: We can't add a `console.error` here because the context will be undefined when you are not in upsell mode. So all the users with the feature properly enabled will see the `console.error` added here

**Pull request merged**
